---
title: "regutools: an R package for the extraction of gene regulatory networks from RegulonDB"
author: "Jose Alquicira Hernandez, Joselyn Chavez, Carmina Barberena Jonas, Jesus Emiliano Sotelo Fonseca, Alejandro Reyes"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{regutools: an R package for data extraction from RegulonDB}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


# Overview

*Escherichia coli K-12 (E-coli)* is the best bacterial organism studied to date. Thousands of papers have been published using *E-coli* as a model system asking how genes are regulated. The throughput of these experiments range from single-gene studies to whole-genome approaches. Twenty years ago, the database *RegulonDB* started collecting, curating and organizing this information into a centralized resource. Thanks to this huge efforts, researchers have had an easy way to access all these data in a database, facilitating the advancements of complete fields, such as *systems biology*. 

The analysis of high-thoughput experiments -such as RNA-seq or ChIP-seq- often requires the integration of databases such as *RegulonDB* in order to give biological interpretations to these data. The *regutools* package is designed to facilitate such integration by providing programatic access to *RegulonDB* within the R environment. The package retreives information from the *RegulonDB* database into *Bioconductor* objects, ready for downstream analyses. 

The package defines the object **regulondb**, which is a data structure that contains the path to a *SQLite* database retrieved from *RegulonDB* along with metadata such as database version and reference genome. The function `download_database()` will retrieve the latest version of the database. The *regutools* package contains functions with the most popular queries to *regutools*, such as retrieving information of which gene targets are regulated by a transcription factor. But users can also design queries that are specific to their analyses. This vignette describes how to use the provided functions and how to design programmatic queries to *regutools*. The general syntax of the function calls of this package is `result <- functionCalled( regulondb, arguments )`. 

# The **regulondb** object

The **regulondb** object is an extension of an **SQLiteConnection** class that host a connection to a database with the table structure defined in the *RegulonDb* database. It contains additional slots that especify the organism, genome version and database version. The function `regulondb()` is the constructor function of **regulondb** objects. This function receives as input a file path to the database file as well as information about the annotation as character vectors.

```{r echo = TRUE}
library(regutools)

## to be replaced with AnnotationHub call ##
download_database(tempdir(), overwrite=FALSE)

ecoli_regulondb <- 
  regulondb(
    database_path = file.path( tempdir(), "regulondb_sqlite3.db" ),
    organism="E.coli", database_version="version 1", 
    genome_version="version 1" )

ecoli_regulondb

```

In order to get an overview of the tables present in a *regulondb* object, we can use the function `list_datasets()`. This function will output all the available tables (datasets) that can be used to build queries. 

```{r echo = TRUE}
list_datasets( ecoli_regulondb )
```

For each table in the database, users can explore the fields (or attributes) of each table using the function `list_attributes`.

```{r}
head( list_attributes( ecoli_regulondb, "GENE" ), 8 )
```

# Retrieving data

Since the *regulondb* object is an extension of the *SQLiteConnection*, users can retreive data from a *regulondb* object using the function `dbGetQuery()`. Additionally, this package provides a wrapper function to build queries to the database. This function is called `get_dataset()` and has a very similar syntax to the `getBM()` function from the biomaRt package. The main arguments of the `get_dataset()` function are a *regulondb* object, a dataset (or table) of the database, the fields of the dataset to retrieve (attributes) and filters to specify what information to get. The code below shows an example where three attributes from the dataset **"GENE"** for the genes *araC*, *crp* and *lacI*. Note that if the `filters=` parameter is empty, the function will retreive all the hits it find in the database.  

```{r echo = TRUE}

get_dataset( regulondb=ecoli_regulondb, 
            dataset="GENE",
            attributes=c("posleft", "posright", "strand", "name"), 
            filters=list("name"=c("araC","crp","lacI")) )

```

Some of the filters, such as *posright* or *posleft*, can be filtered by specifying intervals. For example, the code below indicates that all the start positions *posright" should be between position 1 and position 5000 of the genome. The parameter `inverval=` is used to specify that the filter for that field will be defined by an interval rather than a exact match.

```{r echo = TRUE}
get_dataset( ecoli_regulondb, attributes=c("posright","name"),
        filters=list( "posright"=c(1, 5000) ),
        interval = "posright",
        dataset="GENE" )
```

# The **regulondb_result** object and integration into the *BioC* ecosystem

By default, the function `get_dataset()` outputs a **regulondb_result** object, which is an extension of a **DataFrame** that stores information about the query used to generate this object. This additional information includes the organism name, the database and genome versions, and the table (or dataset) of the *regulondb* object that was queried by `get_dataset()`.

```{r}
res <- get_dataset( regulondb=ecoli_regulondb, 
            dataset="GENE",
            attributes=c("posleft", "posright", "strand", "name"), 
            filters=list("name"=c("araC","crp","lacI")) )
slotNames(res)
```

To enable integration with other Bioconductor packages, we provide the function `convert_to_granges()` which converts a **regulondb_result** object into a **GRanges** object whenever possible. For example, the result stored in the variable `res` has genomic coordinates and it is thus possible convert `res` into a **GRanges** object. 

```{r}
convert_to_granges(res)
```

An alternative way to get to the same result is to use the parameter `output_format=` directly in the function `get_dataset()`. 

```{r}
get_dataset( regulondb=ecoli_regulondb, 
            dataset="GENE",
            attributes=c("posleft", "posright", "strand", "name"), 
            filters=list("name"=c("araC","crp","lacI")), 
            output_format="GRanges" )
```

In a similar manner, the function `convert_to_biostrings()` converts **regulondb** objects into objects from the `r BiocStyle::Biocpkg("Biostrings")` package. Possible outputs of `convert_to_biostrings()` are **DNAStringSet** objects if `seq_type="DNA"` or a **BStringSet** if `seq_type="product"`. 

```{r}
res_dnastring <- get_dataset( regulondb=ecoli_regulondb, 
            dataset="GENE",
            attributes=c("posleft", "posright", "strand", "name", "dna_sequence"), 
            filters=list("name"=c("araC","crp","lacI")) )
res_dnastring <- convert_to_biostrings( res_dnastring, seq_type="DNA" )
res_dnastring
mcols( res_dnastring )
```

```{r}
res_prodstring <- get_dataset( regulondb=ecoli_regulondb, 
            dataset="GENE",
            attributes=c("posleft", "posright", "strand", "name", "product_sequence"), 
            filters=list("name"=c("araC","crp","lacI")) )
res_prodstring <- convert_to_biostrings( res_prodstring, seq_type="product" )
res_prodstring
mcols( res_prodstring )
```

As with the **GRanges** output mentioned above, it is possible for the output of `get_dataset()` to be a **DNAStringSet** object by specifying the parameter `output_format="DNAStringSet"` or a **BStringSet** object by specifying `output_format="BStringSet"`. Note that the functions to convert **regulondb_result** objects will throw errors if there is insufficient information for the coersion to occur. For example, we will get an error if we try to convert into a **GRanges** object when genomic coordinates are missing from the **regulondb_result** object.

# Building your own queries

In the *regutools* package, we have implemented features that are commonly used when quering data from databases: filtering results by partial matching, filtering by numeric intervals, and building complex queries.

## Partial matching

The code below illustrates the concept of partial matching, in which by setting the parameter `partialmatch=` to `TRUE`, the query returns all the gene name in which the word *ara* is contained. 

```{r echo = TRUE}
get_dataset( ecoli_regulondb, 
             attributes=c("posright","name"),
             filters=list("name"="ara"),
             partialmatch ="name",
             dataset="GENE" )
```

Note that setting the parameter `partialmatch=` to `FALSE` will only return genes where the name string is identical to *ara*. 

## Filtering by numeric intervals

In addition to partial matching, queries can be filtered by numeric intervals. For example, in the code below, the parameter `interv=` is set to `"posright"`. By doing this assignment, we are specifying that the values for `"posright"` must lie between the values of `posright` specified in the `filter=` parameter. Thus, the result of this query will be genes whose right positions lie between the coordinates 2000 and 4000000. Note that the use of the `interv=` parameter in the code below is equivalent to setting the parameter `output_format=` to `"GRanges"` and further subsetting the *GRanges* object using the function `subsetByOverlaps()`. 

```{r echo = TRUE}

get_dataset(
  ecoli_regulondb, 
  attributes = c("name", "strand", "posright", "product_name"), 
           dataset = "GENE",
           filters = list(posright=c("2000","4000000") ),
           interv="posright" )
```

## Complex filters

The examples so far have considered queries in which the results are filtered according to single fields from tables. In order to build queries with filters from more than one field, several filters names can be passed as a list to the `filters=` argument. Additionally the `and=` argument is used to specify whether the filtering conditions of the result of the query must be satisfied (`and=TRUE`) or if satisfying a single condition is enough (`and=FALSE`). 

For example, the code below extracts the genes where either `name` or `product_name` contain the word *Ara* or *Ara*, respectively, if the gene is in the forward strand or if the right position of the gene is between 2000 and 40000000. 

```{r}
nrow(get_dataset(
  ecoli_regulondb, 
  attributes = c("name", "strand", "posright", "product_name"), 
           dataset = "GENE",
           filters = list(name=c("ARA"),
                          product_name=c("Ara"),
                          strand=c("forward"),
                          posright=c("2000","4000000")
           ),
           and=FALSE,
           partialmatch = c("name", "product_name") ,
           interv="posright" ))
```

The query below, which is identical to the query above except the `and=` is set to `TRUE`, returns the genes where all of the specified conditions are satisfied. 

```{r}
nrow(get_dataset(
  ecoli_regulondb, 
  attributes = c("name", "strand", "posright", "product_name"), 
           dataset = "GENE",
           filters = list(name=c("ARA"),
                          product_name=c("Ara"),
                          strand=c("forward"),
                          posright=c("2000","4000000")
           ),
           and=TRUE,
           partialmatch = c("name", "product_name") ,
           interv="posright" ))
```

# Functions with frequently used queries

The regutools package provides functions that encode the most frequent queries that are submitted to the *RegulonDB* web resource.

## Extracting regulatory networks

One of the most important information that the *RegulonDB* database contains are manually curated regulatory networks. The *regutools* package exports the function `get_gene_regulators()`, which inputs a vector of gene names and outputs information about what are the transcription factors that regulate such genes together with the regulatory effect (activator, repressor or dual).

```{r echo = TRUE}
get_gene_regulators( ecoli_regulondb, c("araC","fis","crp") )
```

Similarly, the `get_regulatory_network()` function retrieves all the regulatory network from a **regulondb** object. By default, the output will be a list of transcription factor-gene pairs, indicating which transcription factor regulates which gene.

```{r echo = TRUE}
head( get_regulatory_network( ecoli_regulondb ) )
```

But users can also set the parameter `type=` to `"GENE-GENE"` or `"TF-GENE"` to retrieve gene-gene regulatory networks or transcription factor-transcription factor regulatory networks, respectively. 

Users can also use the `get_regulatory_summary()` to retrieve a summary of the transcription factor regulated the expression of a set of given genes. The parameter `gene_regulators` can receive either a vector of gene names or the output of a call to the function `get_gene_regulators()`. The resulting output is a *regulondb_result* object in which each row shows a transcription factor, and the columns display information about the number, percentage and regulatory activity exerted to the genes.

```{r}
get_regulatory_summary( ecoli_regulondb, gene_regulators = c("araC", "modB") )
```
