---
title: "regutools: an R package for the extraction of gene regulatory networks from RegulonDB"
author: "Jose Alquicira Hernandez, Joselyn Chavez, Carmina Barberena Jonas, Jesus Emiliano Sotelo Fonseca, Alejandro Reyes"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{regutools: an R package for data extraction from RegulonDB}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


# Overview

*Escherichia coli K-12 (E-coli)* is the best bacterial organism studied to date. Thousands of papers have been published using *E-coli* as a model system asking how genes are regulated. The throughput of these experiments range from single-gene studies to whole-genome approaches. Twenty years ago, the database *RegulonDB* started collecting, curating and organizing this information into a centralized resource. Thanks to this huge efforts, researchers have had an easy way to access all these data in a database, facilitating the advancements of complete fields, such as *systems biology*. 

The analysis of high-thoughput experiments -such as RNA-seq or ChIP-seq- often requires the integration of databases such as *RegulonDB* in order to give biological interpretations to these data. The *regutools* package is designed to facilitate such integration by providing programatic access to *RegulonDB* within the R environment. The package retreives information from the *RegulonDB* database into *Bioconductor* objects, ready for downstream analyses. 

The package defines the object **regulondb**, which is a data structure that contains the path to a *SQLite* database retrieved from *RegulonDB* along with metadata such as database version and reference genome. The function `download_database()` will retrieve the latest version of the database. The *regutools* package contains functions with the most popular queries to *regutools*, such as retrieving information of which gene targets are regulated by a transcription factor. But users can also design queries that are specific to their analyses. This vignette describes how to use the provided functions and how to design programmatic queries to *regutools*. The general syntax of the function calls of this package is `result <- functionCalled( regulondb, arguments )`. 

# The **regulondb** object

The **regulondb** object is an extension of an **SQLiteConnection** class that host a connection to a database with the table structure defined in the *RegulonDb* database. It contains additional slots that especify the organism, genome version and database version. The function `regulondb()` is the constructor function of **regulondb** objects. This function receives as input a file path to the database file as well as information about the annotation as character vectors.

```{r echo = TRUE}
library(regutools)

## to be replaced with AnnotationHub call ##
download_database(getwd(), overwrite=TRUE)

ecoli_regulondb <- 
  regulondb(
    database_path = file.path( getwd(), "regulondb_sqlite3.db" ),
    organism="E.coli", database_version="version 1", 
    genome_version="version 1" )

ecoli_regulondb

```

In order to get an overview of the tables present in a *regulondb* object, we can use the function `list_datasets()`. This function will output all the available tables (datasets) that can be used to build queries. 

```{r echo = TRUE}
list_datasets( ecoli_regulondb )
```

For each table in the database, users can explore the fields (or attributes) of each table using the function `list_attributes`.

```{r}
head( list_attributes( ecoli_regulondb, "GENE" ), 8 )
```

# Retrieving data

Since the *regulondb* object is an extension of the *SQLiteConnection*, users can retreive data from a *regulondb* object using the function `dbGetQuery()`. Additionally, this package provides a wrapper function to build queries to the database. This function is called `get_dataset()` and has a very similar syntax to the `getBM()` function from the biomaRt package. The main arguments of the `get_dataset()` function are a *regulondb* object, a dataset (or table) of the database, the fields of the dataset to retrieve (attributes) and filters to specify what information to get. The code below shows an example where three attributes from the dataset **"GENE"** for the genes *araC*, *crp* and *lacI*. Note that if the `filters=` parameter is empty, the function will retreive all the hits from the database.  

```{r echo = TRUE}

get_dataset( regulondb=ecoli_regulondb, 
            dataset="GENE",
            attributes=c("posleft", "posright", "strand", "name"), 
            filters=list("name"=c("araC","crp","lacI")) )

```

Some of the filters, such as *posright* or *posleft*, can be filtered by specifying intervals. For example, the code below indicates that all the start positions *posright" should be between position 1 and position 5000 of the genome. The parameter `inverval=` is used to specify that the filter for that field will be defined by an interval rather than a exact match.

```{r echo = TRUE}
get_dataset( ecoli_regulondb, attributes=c("posright","name"),
        filters=list( "posright"=c(1, 5000) ),
        interval = "posright",
        dataset="GENE" )
```

# The **regulondb_result** object and integration into the *BioC* ecosystem

By default, the function `get_dataset()` outputs a **regulondb_result** object, which is an extension of a **DataFrame** that stores information about the query used to generate this object. This additional information includes the organism name, the database and genome versions, and the table (or dataset) of the *regulondb* object that was queried by `get_dataset()`.

```{r}
res <- get_dataset( regulondb=ecoli_regulondb, 
            dataset="GENE",
            attributes=c("posleft", "posright", "strand", "name"), 
            filters=list("name"=c("araC","crp","lacI")) )
slotNames(res)
```

To enable integration with other Bioconductor packages, we provide the function `convert_to_granges()` which converts a **regulondb_result** object into a **GRanges** object whenever possible. For example, the result stored in the variable `res` has genomic coordinates and it is thus possible convert `res` into a **GRanges** object. 

```{r}
convert_to_granges(res)
```

An alternative way to get to the same result is to use the parameter `output_format=` directly in the function `get_dataset()`. 

```{r}
get_dataset( regulondb=ecoli_regulondb, 
            dataset="GENE",
            attributes=c("posleft", "posright", "strand", "name"), 
            filters=list("name"=c("araC","crp","lacI")), 
            output_format="GRanges" )
```

In a similar manner, the function `convert_to_biostrings()` converts **regulondb** objects into objects from the `r BiocStyle::Biocpkg("Biostrings")` package. Possible outputs of `convert_to_biostrings()` are **DNAStringSet** objects if `seq_type="DNA"` or a **BStringSet** if `seq_type="product"`. 

```{r}
res_dnastring <- get_dataset( regulondb=ecoli_regulondb, 
            dataset="GENE",
            attributes=c("posleft", "posright", "strand", "name", "dna_sequence"), 
            filters=list("name"=c("araC","crp","lacI")) )
res_dnastring <- convert_to_biostrings( res_dnastring, seq_type="DNA" )
res_dnastring
mcols( res_dnastring )
```

```{r}
res_prodstring <- get_dataset( regulondb=ecoli_regulondb, 
            dataset="GENE",
            attributes=c("posleft", "posright", "strand", "name", "product_sequence"), 
            filters=list("name"=c("araC","crp","lacI")) )
res_prodstring <- convert_to_biostrings( res_prodstring, seq_type="product" )
res_prodstring
mcols( res_prodstring )
```

As with the **GRanges** output mentioned above, it is possible for the output of `get_dataset()` to be a **DNAStringSet** object by specifying the parameter `output_format="DNAStringSet"` or a **BStringSet** object by specifying `output_format="BStringSet"`. Note that the functions to convert **regulondb_result** objects will throw errors if there is insufficient information for the coersion to occur. For example, we will get an error if we try to convert into a **GRanges** and genomic coordinates are missing from the **regulondb_result** object.

## Partial match

```{r echo = TRUE}
get_dataset( ecoli_regulondb, 
             attributes=c("posright","name"),
             filters=list("name"="ara"),
             partialmatch ="name",
             dataset="GENE" )
```

## Complex queries
```{r echo = TRUE}
dim(get_dataset(
  ecoli_regulondb, 
  attributes = c("name", "strand", "posright", "product_name"), 
           dataset = "GENE",
           filters = list(name=c("ARA"),
                          product_name=c("Ara"),
                          strand=c("forward"),
                          posright=c("2000","4000000")
           ),
           and=FALSE,
           partialmatch = c("name", "product_name") ,
           interv="posright" ))
```
## Complex queries
```{r echo = TRUE}
GetAttr(attributes = c("name", "strand", "posright", "product_name"), 
           dataset = "GENE",
           filters = list(name=c("ARA"),
                          product_name=c("Ara"),
                          strand=c("forward"),
                          posright=c("2000","4000000")
           ),
           and=TRUE,
           partialmatch = c("name", "product_name") ,
           interv="posright" )
```

## Gene regulation 
```{r echo = TRUE}
get_gene_regulations( ecoli_regulondb, c("araC","fis","crp") )
```

## Table format
```{r echo = TRUE}
GetGeneRegulation(c("araC","fis","crp"),
                  format="table")
```

## Retrieve whole network
```{r echo = TRUE}
head(GetNetwork())
```

##  TF-TF network
```{r echo = TRUE}
head(GetNetwork(type = "TF-TF"))
```


