---
title: "regutools: an R package for the extraction of gene regulatory networks from RegulonDB"
author: "Jose Alquicira Hernandez, Joselyn Chavez, Carmina Barberena Jonas, Jesus Emiliano Sotelo Fonseca "
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{regutools: an R package for data extraction from RegulonDB}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


# Overview

*Escherichia coli K-12 (E-coli)* is the best bacterial organism studied to date. Thousands of papers have been published using *E-coli* as a model system asking how genes are regulated. The throughput of these experiments range from single-gene studies to whole-genome approaches. Twenty years ago, the database *regulondb* started collecting, curating and organizing this information into a centralized resource. Thanks to this huge efforts, researchers have had an easy way to access all these data in a database, facilitating the advancements of complete fields, such as *systems biology*. 

The analysis of high-thoughput experiments -such as RNA-seq or ChIP-seq- often requires the integration of databases such as *regulondb* in order to give biological interpretations to these data. The *regutools* package is designed to facilitate such integration by providing programatic access to *regulondb* within the R environment. The package retreives information from the *regulondb* database into *Bioconductor* objects, ready for downstream analyses. 

The package defines the object *regulonData*, which is a data structure that contains the path to a *sqlite* database retrieved from *regulondb* along with metadata such as database version and reference genome. The function `download_regulondb()` will retrieve the latest version of the database. The *regulondb* package contains functions with the most popular queries to *regulondb*, such as retrieving information of which gene targets are regulated by a transcription factor. But users can also design queries that are specific to their analyses. This vignette describes how to use the provided functions and how to design programmatic queries to *regulondb*. The general format of the function calls of this package is `result <- functionCalled( regulonData, arguments )`. 

# The regulonData object

In order to get an overview of the tables present in a *regulonData* object, we can use the function `ListDataSets()`. This function will output all the available tables that can be used to design queries. 

```{r echo = TRUE}
library(regutools)
## to be replaced with AnnotationHub call ##

download_database(getwd())
regulonData <- dbConnect(SQLite(), file.path(getwd(), "regulondb_sqlite3.db"))
class(regulonData)
###
list_datasets()
```

For each table in the database, users can explore the columns of each table using the function `columns`. 

```{r echo = TRUE}
head(ListAttributes("GENE",
               comments = TRUE))
```

# Retrieving data

```{r echo = TRUE}
head(GetAttr(attributes=c("posright","name"),
        dataset="GENE"))
```

## Using filters
```{r echo = TRUE}
GetAttr(attributes=c("posright","name"),
        filters=list("name"=c("araC","crp","lacI")),
        dataset="GENE")
```

## Interval filter
```{r echo = TRUE}
GetAttr(attributes=c("posright","name"),
        filters=list("posright"=c(1,5000)),
        interv=c("posright"),
        dataset="GENE")
```

## Partial match
```{r echo = TRUE}
head(GetAttr(attributes=c("posright","name"),
        filters=list("name"="ara"),
        partialmatch="name",
        dataset="GENE"))
```

## Complex queries
```{r echo = TRUE}
dim(GetAttr(attributes = c("name", "strand", "posright", "product_name"), 
           dataset = "GENE",
           filters = list(name=c("ARA"),
                          product_name=c("Ara"),
                          strand=c("forward"),
                          posright=c("2000","4000000")
           ),
           and=FALSE,
           partialmatch = c("name", "product_name") ,
           interv="posright" ))
```
## Complex queries
```{r echo = TRUE}
GetAttr(attributes = c("name", "strand", "posright", "product_name"), 
           dataset = "GENE",
           filters = list(name=c("ARA"),
                          product_name=c("Ara"),
                          strand=c("forward"),
                          posright=c("2000","4000000")
           ),
           and=TRUE,
           partialmatch = c("name", "product_name") ,
           interv="posright" )
```

## Gene regulation 
```{r echo = TRUE}
GetGeneRegulation(c("araC","fis","crp"))
```

## Table format
```{r echo = TRUE}
GetGeneRegulation(c("araC","fis","crp"),
                  format="table")
```

## Retrieve whole network
```{r echo = TRUE}
head(GetNetwork())
```

##  TF-TF network
```{r echo = TRUE}
head(GetNetwork(type = "TF-TF"))
```

## Formulating biologial questions
Â¿Does the gene "celA" exists?
```{r echo = TRUE, error = TRUE}
head(GetAttr(attributes=c("name"),
             filters=list("name"="celA"),
             dataset="GENE"))
```

```{r echo = TRUE}
head(GetAttr(attributes=c("name","synonyms"),
             filters=list("synonyms"="celA"),
             partialmatch = "synonyms",
             dataset="GENE"))
```
